HINT: Substitute generic parameters in Parameter info

Kobzol opened the pull request:
I'm not sure about using types vs type references for function/method calls.

Fixes: https://github.com/intellij-rust/intellij-rust/issues/4308

changelog: Substitute generic parameters in `Parameter Info`.

ortem commented:
@Kobzol Could you please rebase onto master?

ortem commented on src/main/kotlin/org/rust/ide/utils/CallInfo.kt:
```diff
@@ -22,14 +30,33 @@ class CallInfo private constructor(
-                return CallInfo(ty, parameterNames)
+
+            return when {
+                fn is RsFunction -> {
```
Consider extracting branches bodies into private functions. Maybe something like this (if I understand the whole logic right):
```Kotlin
return when {
    fn is RsFunction -> handleFunction(fn, ty)
    ty is TyFunction -> handleEnum(fn, ty)
    else -> null
}
```

ortem commented on src/main/kotlin/org/rust/ide/utils/CallInfo.kt:
```diff
@@ -38,11 +65,25 @@ class CallInfo private constructor(
         fun resolve(methodCall: RsMethodCall): CallInfo? {
-            return (methodCall.reference.resolve() as? RsFunction)?.let { CallInfo(it) }
+            val function = (methodCall.reference.resolve() as? RsFunction) ?: return null
+            val substitution = let {
```
Probably I'm missing something, could you please explain why we cannot use something like `(methodCall.parentDotExpr.type as? TyFunction)?.paramTypes` as the actual types instead of these substitutions?

Kobzol replied:
`methodCall.parentDotExpr.type` is the type of the return value of the method call, not of the method (function) itself. I haven't found any easier way how to get the actual substituted method type with generics resolved.

ortem commented on src/test/kotlin/org/rust/ide/hints/parameter/RsParameterInfoHandlerTest.kt:
```diff
@@ -166,4 +169,25 @@ class RsParameterInfoHandlerTest
+            let s = S::<i32>();
+            s.foo(/*caret*/);
+        }
+    """, "x: i32", 0)
```
Please add more tests with unknown types, lifetimes, etc.

ortem reviewed (changes requested):
I think it's better to show a generic type name in case of unknown actual types.

Before:
<img width="740" alt="Screenshot 2021-04-08 at 20 10 22" src="https://user-images.githubusercontent.com/4854600/114068386-8a774f00-98a6-11eb-99f2-f2693d102499.png">
After:
<img width="740" alt="Screenshot 2021-04-08 at 20 10 08" src="https://user-images.githubusercontent.com/4854600/114068379-89462200-98a6-11eb-9e8b-a24469bc5c90.png">

The same with lifetimes. BTW do you think lifetimes should be shown? Or maybe just ignore them?

Kobzol commented:
> I think it's better to show a generic type name in case of unknown actual types.

I don't like that too much, as a user I would probably want to see that I'm trying to pass an unresolved/invalid type to the function, and showing `<unknown>` is way more explicit in that regard than showing the original generic parameter name.

Also, I'm not sure how to properly implement it, since e.g. for functions I get the resolved type parameters, where the generic type is already resolved to `TyUnknown`. Maybe zip the resolved types with the original signature and take the original text of the  parameter if the resolved type is unknown? Seems a bit hacky, but it might work.

I would show lifetimes.

Kobzol commented:
@ortem Rebased on master. I added lifetime rendering and more tests.

ortem commented:
> as a user I would probably want to see that I'm trying to pass an unresolved/invalid type to the function, and showing <unknown> is way more explicit in that regard

@Kobzol That's right, I agree with you about being explicit when the types are unresolved or invalid. But I also think in some cases a generic type name might be more informative than `<unknown>`, especially in incomplete code.

Let's check another example.
```Rust
fn add_element<T>(xs: &mut Vec<T>, x: T) { xs.push(x) }

fn main() {
    // ...
    add_element(/*caret*/); // user just started thinking about what to pass in this function
}
```
Before:
<img width="571" alt="Screenshot 2021-07-06 at 15 33 40" src="https://user-images.githubusercontent.com/4854600/124609027-8c35bc00-de6f-11eb-95d5-59a4f71d499e.png">
After:
<img width="562" alt="Screenshot 2021-07-06 at 15 29 42" src="https://user-images.githubusercontent.com/4854600/124608918-71fbde00-de6f-11eb-9634-a8bbb01662b1.png">

Here generic type names indicate that elements of `xs` have the same type as `x`. But this fact is not so clear with `<unknown>` in place. 

What do you think about these cases?

Kobzol commented:
Hmm. Your example convinced me that showing the parameter might be a good idea :) So let's show the actual type if it's not `TyUnknown`, and otherwise show the type from the original parameter declaration?

Kobzol commented:
I have reworked the implementation to be more unified and added proper support for closures. If the type is unknown, the original type parameter name will be used.

Kobzol commented on src/main/kotlin/org/rust/ide/utils/CallInfo.kt:
```diff
@@ -22,27 +30,79 @@ class CallInfo private constructor(
+        private fun buildFunctionParameters(function: RsFunction, ty: Ty): CallInfo {
+            val types = let {
+                val types = (ty as? TyFunction)?.paramTypes
+                    ?: function.valueParameters.mapNotNull { it.typeReference?.type }
```
`self` is actually filled in another place (https://github.com/intellij-rust/intellij-rust/pull/6664/files#diff-59b6929890a0d4ca48aad7ebe026cff696f8b855e7a3ef42076b819648b1caf8R107), so it works (there are several tests that work with `self`, even with UFCS).

The parameters handle this kind of automatically, since we access `valueParameters`, which doesn't contain `self`. But `TyFunction.paramTypes` does (check e.g. `RsParameterInfoHandlerTest::test method`), so we need to drop the self type from it. It wasn't needed before this PR, because the actual parameter types were not used, just `valueParameters` were used before.

But what is wrong is that we mix `paramTypes` with `valueParameters` here and then drop the first one, but if `valueParameters` would be used, that would indeed result in a skipped parameter. Since `ty` is always `TyFunction`, we can solve that by just checking that `ty` is `TyFunction` soon (in `resolve` and then this problem should go away). I pushed the corresponding change.

Kobzol commented:
Hmm, now there's another problem though. The `CallInfo` is used for both normal parameter hints and for inlay hints. With these changes, inlay hints no longer get the name of the corresponding parameter, since the type is known.

The problem is that (for displaying generic parameter types), normal hints use the `type` attribute of the parameter, while inlay hints use the `name` attribute of the parameter. Maybe we should change `CallInfo` so that the parameter contains
1) Pattern
2) Type
3) "Name" (of a generic parameter), which would be used as a hint for inlay parameters and also as a placeholder for unknown types in normal parameter hints?

ortem commented:
@Kobzol Do you mean such hints for enum variants? Seems like only tests about enum variants are failed in `RsInlayParameterHintsProviderTest`.

Kobzol commented:
Well, it should also be the same for tuple structs I guess, but yeah, the enum variants tests are failed, because they use the `CallInfo` in a different way than normal hints.

ortem commented:
OK, this really should be fixed somehow.

Kobzol commented:
I changed how `CallInfo` works. Now inlay hints get the same behaviour as before and the parameter provider should also work as intended. Also now if there is no pattern for the parameter (happens for enum variants or tuple structs), the pattern will not be rendered at all (not even as `_`), it just produced visual clutter.

ortem commented on src/main/kotlin/org/rust/ide/utils/CallInfo.kt:
```diff
@@ -5,44 +5,105 @@
+                element is RsStructItem && element.isTupleStruct ->
+                    element.tupleFields?.tupleFieldDeclList?.map { null to it.typeReference }
+                element is RsPatBinding -> {
+                    val decl = element.parentOfType<RsLetDecl>() ?: return null
```
[minor]
```suggestion
                    val decl = element.ancestorStrict<RsLetDecl>() ?: return null
```

ortem commented on src/main/kotlin/org/rust/ide/utils/CallInfo.kt:
```diff
@@ -5,44 +5,105 @@
 
         fun resolve(methodCall: RsMethodCall): CallInfo? {
-            return (methodCall.reference.resolve() as? RsFunction)?.let { CallInfo(it) }
+            val function = (methodCall.reference.resolve() as? RsFunction) ?: return null
```
[minor] Let's move this method up to group two public `resolve` methods together for better readability

ortem commented on src/main/kotlin/org/rust/ide/utils/CallInfo.kt:
```diff
@@ -5,44 +5,105 @@
+        }
+
+        private fun buildFunctionParameters(function: RsFunction, ty: TyFunction): CallInfo {
+            val types = let {
```
[minor] Executing a block of statements is usually implemented using `run`
```suggestion
            val types = run {
```

ortem reviewed (approved):
LGTM except for minor comments. @Kobzol Thank you!

Kobzol commented:
bors r=ortem

Kobzol commented:
bors retry
